File: Frontend/src/types/stock.ts
// Mirrors the backend API shape exactly (app/Schemas/stock.py) — snake_case, no case-mapping layer.
import type { Destination } from './destinations';

export type UnitCategory = 'vehicle' | 'equipment';
export type SteeringPosition = 'LHD' | 'RHD';
export type AuctionGrade = '5' | '4.5' | '4' | '3.5' | '3' | 'R' | 'RA';
export type UnitStatus = 'in_stock' | 'sold' | 'sourcing';
export type PhotoType = 'exterior' | 'interior' | 'engine_bay' | 'undercarriage' | 'odometer' | 'other';
export type FeatureCategory = 'comfort' | 'safety' | 'exterior' | 'mechanical' | 'equipment_attachment';

export interface UnitImage {
  id: number;
  url: string;
  photo_type: PhotoType;
  alt_text?: string | null;
  sort_order: number;
}

export interface Feature {
  id: number;
  name: string;
  category: FeatureCategory;
}

// UnitSummaryResponse
export interface UnitSummary {
  id: number;
  slug: string;
  category: UnitCategory;
  body_type: string;
  make: string;
  model: string;
  year: number;
  color?: string | null;
  price_usd: number;
  port: string;
  mileage_km?: number | null;
  operating_hours?: number | null;
  auction_grade: AuctionGrade;
  status: UnitStatus;
  steering_position?: SteeringPosition | null;
  transmission?: string | null;
  fuel_type?: string | null;
  created_at: string;
  thumbnail_url?: string | null;
}

// UnitResponse
export interface Unit extends UnitSummary {
  repair_history: boolean;
  one_owner?: boolean | null;
  auction_sheet_url?: string | null;
  chassis_number: string;
  engine?: string | null;
  displacement_cc?: number | null;
  drivetrain?: string | null;
  description: string;
  updated_at: string;
  images: UnitImage[];
  features: Feature[];
}

// UnitPriceUpdate (app/Schemas/stock.py) — the only field PATCH /admin/stock/{id}/price accepts.
export interface UnitPriceUpdate {
  price_usd: number;
}

export interface StockSearchParams {
  category?: UnitCategory;
  body_type?: string;
  make?: string;
  model?: string;
  year_min?: number;
  year_max?: number;
  price_min?: number;
  price_max?: number;
  mileage_max_km?: number;
  auction_grade_min?: AuctionGrade;
  steering_position?: SteeringPosition;
  fuel_type?: string;
  transmission?: string;
  keyword?: string;
  sort?: 'newest' | 'price_asc' | 'price_desc' | 'year_desc' | 'mileage_asc' | 'grade_desc';
  cursor?: number;
  /** Sort-column value of the last unit on the previous page (non-newest sorts). */
  cursor_value?: string;
  limit?: number;
}

export interface StockListResponse {
  items: UnitSummary[];
  next_cursor?: number | null;
  next_cursor_value?: string | null;
  /** Matching units; only present on the first page. */
  total?: number | null;
}

// ---- GET /stock/facets (StockFacetsResponse) ---------------------------------

export interface FacetCount {
  value: string;
  count: number;
}

export interface StockFacets {
  total: number;
  vehicles: number;
  equipment: number;
  /** Top makes by in-stock count, capped server-side. */
  makes: FacetCount[];
  /** Top body types by in-stock count, capped server-side. */
  body_types: FacetCount[];
  /** Only values present in stock (at most LHD/RHD). */
  steering_positions: FacetCount[];
  fuel_types: FacetCount[];
  grades: FacetCount[];
  year_min?: number | null;
  year_max?: number | null;
  price_min?: number | null;
  price_max?: number | null;
}

// ---- GET /stock/{slug}/insights (UnitInsightsResponse) ----------------------

export interface GradeCount {
  grade: AuctionGrade;
  count: number;
}

export interface PricePoint {
  id: number;
  slug: string;
  year: number;
  price_usd: number;
  /** mileage_km for vehicles, operating_hours for equipment — see MarketPosition.usage_unit */
  usage?: number | null;
  auction_grade: AuctionGrade;
  is_current: boolean;
}

export type MarketScope = 'model' | 'body_type' | 'category';

export interface MarketPosition {
  scope: MarketScope;
  label: string;
  peer_count: number;
  price_min?: number | null;
  price_median?: number | null;
  price_max?: number | null;
  price_avg?: number | null;
  usage_avg?: number | null;
  usage_unit: 'km' | 'hrs';
  /** % of peers priced at or below this unit (0–100) */
  price_percentile?: number | null;
  /** % of peers with usage at or below this unit (0–100) */
  usage_percentile?: number | null;
  grade_distribution: GradeCount[];
}

export interface UnitInsights {
  market: MarketPosition;
  comparables: UnitSummary[];
  price_points: PricePoint[];
  destinations: Destination[];
  feature_catalog: Feature[];
}

File: Backend/app/Schemas/stock.py
# app/Schemas/stock.py
# Unit request/response schemas (databaseschema.md §2).
from datetime import datetime
from typing import List, Optional

from pydantic import BaseModel, ConfigDict, Field

from app.Schemas.destination import DestinationResponse


class UnitImageResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    url: str
    photo_type: str
    alt_text: Optional[str] = None
    sort_order: int


class FeatureResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    name: str
    category: str


class UnitSummaryResponse(BaseModel):
    """List/search row — the columns a stock grid actually renders."""
    model_config = ConfigDict(from_attributes=True)

    id: int
    slug: str
    category: str
    body_type: str
    make: str
    model: str
    year: int
    color: Optional[str] = None
    price_usd: float
    port: str
    mileage_km: Optional[int] = None
    operating_hours: Optional[int] = None
    auction_grade: str
    status: str
    steering_position: Optional[str] = None
    transmission: Optional[str] = None
    fuel_type: Optional[str] = None
    created_at: datetime
    thumbnail_url: Optional[str] = None


class UnitResponse(UnitSummaryResponse):
    """Full detail view — everything UnitSummaryResponse has, plus the rest of the row and its two eager-loaded collections."""
    repair_history: bool
    one_owner: Optional[bool] = None
    auction_sheet_url: Optional[str] = None
    chassis_number: str
    engine: Optional[str] = None
    displacement_cc: Optional[int] = None
    drivetrain: Optional[str] = None
    fuel_type: Optional[str] = None
    description: str
    updated_at: datetime

    images: List[UnitImageResponse] = []
    features: List[FeatureResponse] = []


class StockSearchParams(BaseModel):
    """Query params for GET /stock — every field is optional; an unset field applies no filter."""

    category: Optional[str] = None
    body_type: Optional[str] = None
    make: Optional[str] = None
    model: Optional[str] = None
    year_min: Optional[int] = None
    year_max: Optional[int] = None
    price_min: Optional[float] = None
    price_max: Optional[float] = None
    mileage_max_km: Optional[int] = None
    auction_grade_min: Optional[str] = None
    steering_position: Optional[str] = None
    fuel_type: Optional[str] = None
    transmission: Optional[str] = None
    keyword: Optional[str] = Field(default=None, description="Free-text search over make/model/description")

    sort: str = Field(default="newest", pattern="^(newest|price_asc|price_desc|year_desc|mileage_asc|grade_desc)$")
    cursor: Optional[int] = Field(default=None, description="Last-seen unit id from the previous page")
    cursor_value: Optional[str] = Field(default=None, description="Sort-column value of that last unit (non-newest sorts)")
    limit: int = Field(default=24, ge=1, le=100)


class StockListResponse(BaseModel):
    items: List[UnitSummaryResponse]
    next_cursor: Optional[int] = None
    next_cursor_value: Optional[str] = None
    total: Optional[int] = Field(default=None, description="Matching units; only computed on the first page")


class FacetCount(BaseModel):
    value: str
    count: int


class StockFacetsResponse(BaseModel):
    """Aggregate counts over in-stock, non-deleted units — what the storefront's browse-by-make / body-type tiles, the hero finder, and the category landing…"""

    total: int
    vehicles: int
    equipment: int
    makes: List[FacetCount]
    body_types: List[FacetCount]
    steering_positions: List[FacetCount] = []
    fuel_types: List[FacetCount] = []
    grades: List[FacetCount] = []
    year_min: Optional[int] = None
    year_max: Optional[int] = None
    price_min: Optional[float] = None
    price_max: Optional[float] = None


class StockCountResponse(BaseModel):
    count: int


class UnitPriceUpdate(BaseModel):
    price_usd: float


# ---- Detail-page insights (GET /stock/{slug}/insights) ------------------------- Everything the storefront's "market position" / shipping / equipment…

class GradeCount(BaseModel):
    grade: str
    count: int


class PricePoint(BaseModel):
    """One dot on the price-vs-usage scatter."""
    id: int
    slug: str
    year: int
    price_usd: float
    usage: Optional[int] = None
    auction_grade: str
    is_current: bool = False


class MarketPosition(BaseModel):
    scope: str = Field(description="'model' (same make+model), 'body_type', or 'category' — widest scope that had enough peers")
    label: str
    peer_count: int
    price_min: Optional[float] = None
    price_median: Optional[float] = None
    price_max: Optional[float] = None
    price_avg: Optional[float] = None
    usage_avg: Optional[int] = None
    usage_unit: str = "km"
    price_percentile: Optional[int] = Field(default=None, description="% of peers priced at or below this unit (0-100)")
    usage_percentile: Optional[int] = Field(default=None, description="% of peers with usage at or below this unit (0-100)")
    grade_distribution: List[GradeCount] = []


class UnitInsightsResponse(BaseModel):
    market: MarketPosition
    comparables: List[UnitSummaryResponse] = []
    price_points: List[PricePoint] = []
    destinations: List[DestinationResponse] = []
    feature_catalog: List[FeatureResponse] = []
